### PR TITLE
Support multi-GPU `cupy.cuda.cufft.Plan1d`

### PR DESCRIPTION
**Pull request by leofang**
***--- PR description updated, please read ---*** 

This PR adds initial support for multi-GPU cuFFT. I only target `Plan1d` here because it's easier to test and to understand the corner cases. This PR makes the necessary preparation for future extension. For example, I will support `PlanNd` in the next PR.

In short, this PR allows multi-GPU cuFFT plans to be used with both high- and low-level API calls:
```python
import cupy as cp


a = cp.random.random(1024).astype(cp.complex64)
a += 1j*cp.random.random(1024)

cp.fft.config.use_multi_gpus = True  # flag added in this PR
cp.fft.config.set_cufft_gpus(2)      # handle added in this PR
out = cp.fft.fft(a)                  # this just works
```
```python
import cupy as cp


batch = 7
length = 512

# use 2 GPUs
a = cp.random.random((batch, length)) +1j*cp.random.random((batch, length))
a = a.astype(cp.complex64)
plan = cp.cuda.cufft.Plan1d(length, cp.cuda.cufft.CUFFT_C2C, batch,
                            use_multi_gpus=True, devices=2)
out = plan.get_output_array(a)
plan.fft(a, out, cp.cuda.cufft.CUFFT_FORWARD)

# use 1 GPU
out2 = cp.fft.fft(a)

# results should be identical 
assert cp.allclose(out, out2)
```
The second example is nice because `Plan1d` can actually be retrieved by `cupyx.scipy.fftpack.get_fft_plan()`.

Multi-GPU cuFFT are beneficial for at least two reasons:
1. Bypassing memory limitation: every GPU has limited memory, and even worse in order to run `cupy.fft.fft()`, for example, it would need additional buffer space. So it turns out one cannot even FFT a 8G array on a P100 with 16G memory. Having extra workhorses would alleviate this limitation.
2. Reasonable performance: while [it's not guaranteed to be faster using multiple GPUs](https://docs.nvidia.com/cuda/cufft/index.html#multiple-GPU-cufft-transforms), for users who have NVLink interconnects, the bandwidth is large enough that performance is acceptable for a huge array. It seems that NVLinks can support more simultaneous async transfers than allowed by the engine counts (which shows up in `nvprof`/`nvvp` as `Memcpy PtoP` --- on a DGX-2 I found overlapped memcpy among 16 GPUs is possible with this PR!)

Under the hood, the `cupy.cuda.cufft` module manages all nuisances. In particular, these two cuFFT APIs are *not* used: `cufftXtMalloc()` and `cufftXtFree()`. This decision is made because 
1. we don't want memory allocation outside of CuPy memory pool 
2. we want to reuse the buffers when the plan is pre-allocated (as shown above) and reused

So some efforts are invested to ensure this really works. This approach is encouraged by NVIDIA architect @maxpkatz, who is very kind to offer help ❤️ This PR would have been much more difficult without his support! Thank you (again), Max!

(Disclaimer: Max did not review nor approve this PR. He only provided technical discussion and consultation.)

Also note that when using `Plan1d.fft()` directly, passing NumPy arrays are allowed. I prefer to keep this feature undocumented and polish it later (likely in the next PR, during which I will need to re-evaluate this feature again for 2D/3D FFTs). Is it OK?

TODO:
- [x] more local tests and polishes
- [x] avoid `cufftXtMalloc` and `cufftXtFree`
- [x] make it work with high-level API (`cupy.fft.fft()`)
- [x] write test suites ~~and docs~~
- [x] fix read the docs

### CODE REVIEW

File: cupy/fft/fft.py [hunk 1]
```diff
@@ -172,6 +178,12 @@ def _fft(a, s, axes, norm, direction, value_type='C2C', overwrite_x=False,
         a = _exec_fft(a, direction, value_type, norm, axes[-1], overwrite_x,
                       out_size)
 
+    # TODO(leofang): ask for help
+    # For some reason, the assignment operator in _fft_c2c() (a = _exec_fft(
+    # ...)) silently switched the current device if multi-GPU cuFFT is used.
+    # This hotfix ensures we use whichever device we started with.
+    if config.use_multi_gpus:
+        cupy.cuda.runtime.setDevice(a.data.device_id)
```
Comment (leofang) [inline cupy/fft/fft.py]: For reviewer(s): Please note this observation. I can't understand why would the assignment operation in the Python space touched the CUDA device setting...

Comment (asi1024) [reply cupy/fft/fft.py]: Does it cause also in `master` branch, or only in this branch?

Comment (leofang) [reply cupy/fft/fft.py]: Only here as far as I can tell...

Comment (leofang) [reply cupy/fft/fft.py]: Using multiple ways, I verified that this device switch did not happen in the Python space. Rather, it happened in the last stage of the multi-GPU cuFFT plan execution. The last device provided to `cufftXtSetGPUs()` was set enabled. So, this is a bug in cuFFT, not in this PR.

Comment (leofang) [reply cupy/fft/fft.py]: Sorry for the noise. I removed an earlier comment made when the diagnosis wasn't fully completed, but the conclusion stands: ***This device switching bug is in cuFFT, not in this PR.*** 

With NVIDIA architect @maxpkatz's help, we confirm that when using multiple GPUs, after `cufftDestroy()` completes the current device is silently switched to the last one provided to `cufftXtSetGPUs()`. 

As a workaround, we can either use the hotfix here, or just cache the current device when entering `Plan1d.fft()`. I prefer the latter, what do you think @asi1024?

Comment (leofang) [reply cupy/fft/fft.py]: The internal NVIDIA ticket number filed by @maxpkatz is 2761341.

**Comment by leofang** [comment]
@asi1024 If the team is interested in supporting multi-GPUs I’ll start writing some tests. Docs can wait for another PR — this is the ground work for supporting multi-GPU `PlanNd`, which is my actual interest 🙂

**Comment by leofang** [comment]
@asi1024 I took some time on writing tests. The main PR description is fully rewritten. PTAL. Thanks!

**Comment by leofang** [comment]
Tagging people who might be interested? @grlee77 @dxm447

**Comment by grlee77** [comment]
Haven't had a chance to take a look yet but I am very interested. Thanks for working on this!

**Comment by dxm447** [comment]
Thanks! I will look into it @leofang

File: cupy/fft/config.py [hunk 2]
```diff
@@ -1,3 +1,23 @@
-
 global enable_nd_planning
```
Comment (leofang) [inline cupy/fft/config.py]: @grlee77 quick question: Do you recall why this needs to be `global`? Wouldn't normal module-level variables just work?

**Comment by leofang** [comment]
Thanks, @dxm447. If you have time, could you please help me do the following test: First set `cupy.fft.config.enable_nd_planning = False` to enforce using 1D plans over all axes, then set `cupy.fft.config.use_multi_gpus` and `cupy.fft.config.set_cufft_gpus()` as shown in the [demo](https://github.com/cupy/cupy/pull/2644#issue-341352589), and finally run a multi-dimensional FFT hopefully on multiple GPUs. 

It should just work; if it does not, it's considered a bug and should be fixed in this PR. And if it works, let me know the performance. It'd not be surprising to see worse performance since data movement is involved. I am just curious and need that as a baseline when preparing and comparing multi-GPU `PlanNd`.

File: cupy/fft/config.py (hunk 2 above)
Comment (grlee77) [reply cupy/fft/config.py]: I'm not sure why I had that there. I don't think it needs to be global.

Comment (leofang) [reply cupy/fft/config.py]: Ok, will remove it. Thanks.

**Comment by leofang** [comment]
> Also note that when using `Plan1d.fft()` directly, passing NumPy arrays are allowed.

In other words, this just works:
```python
import cupy as cp
import numpy as np


shape = (128, 512, 512)
a_np = np.random.random(shape)+1j*np.random.random(shape)
a = cp.asarray(a_np)
cp.fft.config.set_cufft_gpus(2)

print("use multi-gpu + Plan1d...")
cp.fft.config.enable_nd_planning = False
cp.fft.config.use_multi_gpus = True
b = cp.fft.fftn(a)

print("use single gpu + PlanNd...")
cp.fft.config.enable_nd_planning = True
cp.fft.config.use_multi_gpus = False
c = cp.fft.fftn(a)

print("use multi-gpu + Plan1d + numpy...")
cp.fft.config.enable_nd_planning = False
cp.fft.config.use_multi_gpus = True
d = cp.fft.fftn(a_np)

print("use numpy...")
e = np.fft.fftn(a_np)

print("multi-gpu + Plan1d = single gpu + PlanNd?", cp.allclose(b, c))
print("multi-gpu + Plan1d = numpy?", cp.allclose(b, d))
print("single gpu + Plan1d = pure numpy?", cp.allclose(c, d))
print("multi-gpu + Plan1d + numpy = pure numpy?", cp.allclose(d, e))
```

File: cupy/fft/fft.py (hunk 1 above)
Comment (leofang) [reply cupy/fft/fft.py]: I took the liberty and moved the fix to `Plan1d`'s destructor in f3a876e.

File: cupy/cuda/cufft.pyx [hunk 3]
```diff
@@ -95,17 +144,160 @@ cpdef inline check_result(int result):
...
+    cdef XtArray* xtArr = <XtArray*>ptr
+    cdef XtArrayDesc* xtArr_desc = xtArr.descriptor
+    PyMem_Free(xtArr_desc)
+    PyMem_Free(xtArr)
+
+
 class Plan1d(object):
-    def __init__(self, int nx, int fft_type, int batch):
+    def __init__(self, int nx, int fft_type, int batch, *,
+                 use_multi_gpus=False, devices=None, out=None):
```
Comment (asi1024) [inline cupy/cuda/cufft.pyx]: Could you add documents about these arguments?

File: cupy/fft/fft.py [hunk 4]
```diff
@@ -395,22 +402,23 @@ def _fftn(a, s, axes, norm, direction, value_type='C2C', order='A', plan=None,
     # sort the provided axes in ascending order
     axes = tuple(sorted(np.mod(axes, a.ndim)))
 
+    xp = cupy.get_array_module(a)
```
Comment (asi1024) [inline cupy/fft/fft.py]: Why do you want to support for numpy.ndarray inputs? Other CuPy's routines do not accept numpy.ndarray as input basically.

File: tests/cupy_tests/fft_tests/test_fft.py [hunk 5]
```diff
@@ -50,6 +50,42 @@ def test_func(self, *args, **kw):
...
+        def test_func(self, *args, **kw):
+            use_multi_gpus = config.use_multi_gpus
+            _devices = config._devices
+
+            try:
+                for gpus in gpu_configs:
+                    try:
+                        nGPUs = len(gpus)
+                        if nGPUs < 2:
+                            raise ValueError('Must use at least two gpus')
```
Comment (asi1024) [inline tests/cupy_tests/fft_tests/test_fft.py]: Please use `assert len(gpus) >= 2` because `testing.numpy.cupy_*(..., accept_error=ValueError, ...)` cannot distinguish between this `ValueError` and a `ValueError` raised in a CuPy's function.

File: tests/cupy_tests/fft_tests/test_fft.py [hunk 6]
```diff
@@ -127,6 +163,93 @@ def test_ifft(self, xp, dtype):
...
+    @multi_gpu_config(gpu_configs=[[0, 1], [1, 0]])
+    @testing.for_complex_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
+                                 contiguous_check=False)
+    def test_fft(self, xp, dtype):
+        a = testing.shaped_random(self.shape, xp, dtype)
+        out = xp.fft.fft(a, n=self.n, norm=self.norm)
+
+        # np.fft.fft alway returns np.complex128
+        if xp == np and dtype is np.complex64:
```
Comment (asi1024) [inline tests/cupy_tests/fft_tests/test_fft.py]: `xp is np` seems more natural.

File: cupy/cuda/cufft.pyx [hunk 7]
```diff
@@ -95,17 +143,160 @@ cpdef inline check_result(int result):
         raise CuFFTError(result)
 
 
+# for debugging
+def _multi_gpu_check_data_integraty(intptr_t in_arr=0):
```
[the remark this answers is not in the document]
Comment (leofang) [reply cupy/cuda/cufft.pyx]: Ah yes, all debug functions shall be removed before merging.

File: cupy/cuda/cufft.pyx (hunk 3 above)
Comment (leofang) [reply cupy/cuda/cufft.pyx]: I intended to complete the doc in a separate PR, but if you think it's more appropriate to be done here, may I suggest instead of adding a docstring to `Plan1d`, we add to 
- `cupy/docs/source/reference/fft.rst` 
- `cupy.fft.config.set_ cufft_gpus()` (the new interface)
- `cupyx.scipy.fftpack.get_fft_plan()` (it can return a multi-GPU plan now)

This is because `Plan1d` is not officially documented and is a semi-public API (only accessible via `get_fft_plan()`).

File: cupy/fft/fft.py (hunk 4 above)
Comment (leofang) [reply cupy/fft/fft.py]: This is because CuPy's internals take too much device memory. For example, as mentioned in the revised PR description, it's impossible to FFT a 8G array on one P100. One would think 2 P100s would help (which has 32G memory in total), but it turns out not possible either. 

But, starting from the host directly, I can FFT a 12G array without problem on 2 P100s. I can even try a 32G array on 2 V100s if you wish. I think this is the most important advantage of multi-GPU FFT.

> Other CuPy's routines do not accept numpy.ndarray as input basically.

True, but other CuPy routines do not have NVIDIA's native multi-GPU support either 🙂 so whenever there is support, I think we should take advantage of it.

As I said, we could either keep this "NumPy-in, NumPy-out" feature undocumented until I add support of multi-GPU `PlanNd`, or mark it experimental. Either is better than nothing.

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 6 above)
Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py]: That's a good catch. I didn't think carefully when copying. I correct all such occurrences throughout this test suite.

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 5 above)
Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py]: Thank you. I didn't consider this. I fixed it.

File: cupy/cuda/cufft.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (leofang) [reply cupy/cuda/cufft.pyx]: Removed.

**Comment by leofang** [comment]
@asi1024 Thank you for your review! I've addressed a few points and replied to two of your questions (doc & NumPy support). Let me know what you think and I'll make changes.

**Comment by leofang** [comment]
Added docs and resolved merge conflicts.

File: cupy/fft/config.py [hunk 8]
```diff
@@ -1,3 +1,30 @@
-
-global enable_nd_planning
 enable_nd_planning = True
+use_multi_gpus = False
+_devices = None
+
+
+def set_cufft_gpus(gpus):
```
Comment (asi1024) [inline cupy/fft/config.py]: We don't want this API because we want users to declare `Pland1d` explicitly in such a situation.

File: cupy/cuda/cufft.pyx [hunk 9]
```diff
@@ -105,17 +154,133 @@ cpdef size_t getVersion() except? -1:
...
+    cdef XtArray* xtArr = <XtArray*>ptr
+    cdef XtArrayDesc* xtArr_desc = xtArr.descriptor
+    PyMem_Free(xtArr_desc)
+    PyMem_Free(xtArr)
+
+
 class Plan1d(object):
-    def __init__(self, int nx, int fft_type, int batch):
+    def __init__(self, int nx, int fft_type, int batch, *,
+                 use_multi_gpus=False, devices=None, out=None):
```
Comment (asi1024) [inline cupy/cuda/cufft.pyx]: `use_multi_gpus` keyword argument seems redundant.

Comment (asi1024) [reply cupy/cuda/cufft.pyx]: I feel a little confused that the keyword argument `devices` is interpreted as the number of devices if an integer is given, and as the device ids if a tuple is given.
How about taking a single device id or a tuple of device ids?

File: cupy/fft/fft.py (hunk 4 above)
Comment (asi1024) [reply cupy/fft/fft.py]: Could you split this fix to another PR? It is certainly related to this PR, but it is over the scope of multi-GPU support.

Comment (leofang) [reply cupy/fft/fft.py]: Sorry, what fix were you referring to? And if you don't want `set_cufft_gpus()`, anything changed in `cupy/fft/fft.py` is meaningless and should be undone, as high-level APIs need not be touched at all.

File: cupy/fft/config.py (hunk 8 above)
Comment (leofang) [reply cupy/fft/config.py]: May I ask what is the concern here? I thought supporting high-level APIs would be appreciated...

File: cupy/cuda/cufft.pyx (hunk 9 above)
Comment (leofang) [reply cupy/cuda/cufft.pyx]: > `use_multi_gpus` keyword argument seems redundant.

No, it's used to split the code path within `Plan1d`. The required steps to initialize a single-GPU plan is different from those for a multi-GPU one.

> I feel a little confused that the keyword argument `devices` is interpreted as the number of devices if an integer is given, and as the device ids if a tuple is given.

This is to be in line with `cupy.cuda.nccl.NcclCommunicator.initAll()`: https://docs-cupy.chainer.org/en/latest/reference/generated/cupy.cuda.nccl.NcclCommunicator.html#cupy.cuda.nccl.NcclCommunicator.initAll

File: cupy/fft/config.py (hunk 8 above)
Comment (leofang) [reply cupy/fft/config.py]: And I was under the impression that you were on board, since you marked this https://github.com/cupy/cupy/pull/2644#discussion_r350168492 as resolved. To remove this, I will also have to undo all the docs I just wrote 🙁

Comment (leofang) [reply cupy/fft/config.py]: ...and I just realized tests have to be rewritten too, as I didn't test `Plan1d` directly. 😞

**Comment by leofang** [comment]
> Could you add the result of the benchmarks to the description of this PR?

Sure. Rule of thumb: **If 1 GPU works, don't use multiple GPUs.**

For problem sizes that can fit in one GPU (starting with a CuPy array), 2-GPU FFT is always slower by a factor of 4 - 10 due to extra data transfer (the actually FFT itself is fast), and the performance is worse with more GPUs.

For problem sizes that cannot fit in one GPU (starting with an NumPy array directly), I get a linear scaling:
- 24GB array, 2 GPU (with NVLinks) -> ~10s
- 48GB array, 4 GPU (with NVLinks) -> ~20s
- 96GB array, 8 GPU (with NVLinks) -> ~40s

which shows the bottleneck is always in data transfer. Linear scaling is already the best scenario. (This shows the actual computation probably has a good weak/constant scaling!)

File: cupy/cuda/cufft.pyx [hunk 10]
```diff
@@ -168,6 +434,171 @@ class Plan1d(object):
...
+                            b[start:start+count].data, size, curr_stream)
+                        if i != 0:
+                            prev_event = self.events[i-1]
+                            curr_stream.wait_event(prev_event)
+                        curr_event = self.events[i]
+                        curr_event.record(curr_stream)
+                        if i == nGPUs - 1:
+                            curr_event.synchronize()
+                        start += count
+                    assert start == b.size
```
Comment (leofang) [inline cupy/cuda/cufft.pyx]: Given that the multi-GPU FFT itself is blastingly fast, I think this naive "scattering" scheme could be improved, such as using a tree (so that the intermediate nodes can do both copy-in and copy-out at the same time). Too bad NCCL does not have a collective scatter call...

File: cupy/cuda/cufft.pyx [hunk 11]
```diff
@@ -168,6 +434,171 @@ class Plan1d(object):
...
+                            xtArr_buffer[i], size, curr_stream)
+                        if i != 0:
+                            prev_event = self.events[i-1]
+                            curr_stream.wait_event(prev_event)
+                        curr_event = self.events[i]
+                        curr_event.record(curr_stream)
+                        if i == nGPUs - 1:
+                            curr_event.synchronize()
+                        start += count
+                    assert start == b.size
```
Comment (leofang) [inline cupy/cuda/cufft.pyx]: Ditto.

File: cupy/fft/fft.py (hunk 1 above)
Comment (asi1024) [reply cupy/fft/fft.py]: I feel that caching the current device in `Plan1d.__del__()` seems better too. Thanks for your investigation!

Comment (asi1024) [reply cupy/fft/fft.py]: These lines may be fixed after the ticket is resolved. Could you leave this comment as a TODO?
https://github.com/cupy/cupy/pull/2644/files#diff-0539f58816d3895bb08f76961618117eR391-R395

File: cupy/fft/fft.py (hunk 4 above)
Comment (asi1024) [reply cupy/fft/fft.py]: Sorry for the confusion. The above comment refers to `numpy.ndarray.` support.
How about supporting multi-GPU `Plan1d` in this PR, and `numpy.ndarray` input in another PR? We need to discuss whether to support numpy inputs, and it will take a long time. So I want to split the discussion into 2 PRs and finish multi-GPU support first.

File: cupy/fft/config.py (hunk 8 above)
Comment (asi1024) [reply cupy/fft/config.py]: I did not doubt these API when I marked the comment as resolved, but now I think that high-level APIs are not necessary for low-level performance improvement. Sorry to ask you for redundant work.

File: cupy/cuda/cufft.pyx (hunk 9 above)
Comment (asi1024) [reply cupy/cuda/cufft.pyx]: The appropriate value of `use_multi_gpus` is uniquely determined from `devices`?

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 6 above)
Comment (asi1024) [reply tests/cupy_tests/fft_tests/test_fft.py]: Could you fix `xp == cupy` too?

File: cupy/fft/fft.py (hunk 4 above)
Comment (leofang) [reply cupy/fft/fft.py]: For the `numpy.ndarray` support, may I suggest instead of completely wiping it out from the codebase, we simply comment it out and leave a TODO?

File: cupy/cuda/cufft.pyx (hunk 9 above)
Comment (leofang) [reply cupy/cuda/cufft.pyx]: I haven't made it so, but we can if you wish. In that case, users just specify `devices`, and with any meaningful `devices`,  the internal `use_multi_gpus` will be set to `True`. What do you think?

File: cupy/fft/fft.py (hunk 1 above)
Comment (leofang) [reply cupy/fft/fft.py]: We no longer need any hotfix in `cupy/fft/fft.py` after the commit f3a876e. Did you mean you want to label the linked comment as a TODO? I can certainly do this, but in the foreseeable future we'll need this hotfix, because the bug is present at least up to CUDA 10.1 Update 2 and very likely also in 10.2 (I don't have the environment to test). Even if NVIDIA fixes it in 10.3 or 11.0, we still need to support all earlier CUDAs until their EOF. So, instead of a TODO, I consider this as "won't fix"...

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 6 above)
Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py]: I don't understand. When `xp is cupy`, the returned type is `complex64`. What do you want to fix?

File: cupy/fft/config.py (hunk 8 above)
Comment (leofang) [reply cupy/fft/config.py]: In this case, I can't reuse all of the existing FFT tests. Do you need a separate test for `Plan1d`?

File: cupy/fft/fft.py (hunk 4 above)
Comment (leofang) [reply cupy/fft/fft.py]: The discussion for NumPy support is in #2742.

File: cupy/fft/fft.py (hunk 1 above)
Comment (asi1024) [reply cupy/fft/fft.py]: > Even if NVIDIA fixes it in 10.3 or 11.0, we still need to support all earlier CUDAs until their EOF.
I see. I agree that we don't need to fix it. Thanks.

File: cupy/fft/fft.py (hunk 4 above)
Comment (leofang) [reply cupy/fft/fft.py]: @asi1024, how about my suggestion https://github.com/cupy/cupy/pull/2644#discussion_r352445583 above?

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 6 above)
Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py]: @asi1024?

**Comment by leofang** [comment]
@asi1024 Could you address my questions above? I am a bit losing track of your expectations here...Thanks.

File: cupy/fft/fft.py (hunk 4 above)
Comment (asi1024) [reply cupy/fft/fft.py]: Do you mean that you will remove the "NumPy-in, NumPy-out" from the current code? If so, I agree with your suggestion.

File: tests/cupy_tests/fft_tests/test_fft.py [hunk 12]
```diff
@@ -438,6 +560,80 @@ def test_fft_error_on_wrong_plan(self, dtype):
...
+@testing.with_requires('numpy>=1.10.0')
+@testing.multi_gpu(2)
+class TestMultiGpuPlanCtxManagerFft(unittest.TestCase):
+    @multi_gpu_config(gpu_configs=[[0, 1], [1, 0]])
+    @testing.for_complex_dtypes()
+    @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
+                                 contiguous_check=False)
+    def test_fft(self, xp, dtype):
+        a = testing.shaped_random(self.shape, xp, dtype)
+        if xp == cupy:
```
Comment (asi1024) [inline tests/cupy_tests/fft_tests/test_fft.py]: Could you fix to `xp is cupy`?

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 6 above)
Comment (asi1024) [reply tests/cupy_tests/fft_tests/test_fft.py]: Sorry to confuse you. I commented it here because it is related to `xp is np`, but you don't need to fix this line anymore. https://github.com/cupy/cupy/pull/2644/files#r358111247

Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py]: Ah, OK. More legacy code to change...Thanks for catching them.

**Comment by leofang** [comment]
@asi1024 I think this is relevant to the discussion in #2743: If you want to expose only the low-level `Plan1d` object for using multi GPUs, and you want to remove all the changes I made to `cupy/fft/config.py` here, it's better to make `cupyx.scipy.fftpack.get_fft_plan()` a stable API so that it can return a multi-GPU plan.

File: cupy/cuda/cufft.pyx (hunk 9 above)
Comment (leofang) [reply cupy/cuda/cufft.pyx]: @asi1024 How about this?

File: cupy/fft/config.py (hunk 8 above)
Comment (leofang) [reply cupy/fft/config.py]: @asi1024 How about my above question for the tests? And do you want everything here to be removed (`use_multi_gpus`, `_devices`, and `set_cufft_gpus`)?

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 12 above)
Comment (leofang) [reply tests/cupy_tests/fft_tests/test_fft.py]: Done.

**Comment by leofang** [comment]
@asi1024 Any chance you have time to address https://github.com/cupy/cupy/pull/2644#discussion_r358294355 and https://github.com/cupy/cupy/pull/2644#discussion_r352446368? 

Another thought: If we are wiping out the high-level APIs for turning on multi-GPU as you suggested, is it then OK to add a `multi_gpu=False` argument to `cupyx.scipy.fftpack.get_fft_plan()` so that it can return a multi-GPU plan if needed? I still don't think it's a good idea to expose the low-level `cupy.cuda.fft` interface, which was considered private API afaik. 

Thanks.

File: cupy/cuda/cufft.pyx (hunk 10 above)
Comment (leofang) [reply cupy/cuda/cufft.pyx]: When the main bottleneck is in data transfer rather than in communication overhead, a tree will be more costly than a plain loop. So for the purpose here we can't do better. Marking this as resolved.

File: cupy/cuda/cufft.pyx (hunk 11 above)
Comment (leofang) [reply cupy/cuda/cufft.pyx]: See https://github.com/cupy/cupy/pull/2644/files#r377135529.

File: cupy/fft/config.py (hunk 8 above)
Comment (asi1024) [reply cupy/fft/config.py]: Sorry, I thought we could get `device` from `plan`, but actually `Plan1d` does not have `device` attribute.

Comment (asi1024) [reply cupy/fft/config.py]: However, this function may be removed in the future. CuPy does not support multiple-GPU computation in one operation currently, but it may be supported in the future, and a feature for multiple-GPU device management may also be supported.

File: cupy/cuda/cufft.pyx (hunk 9 above)
Comment (asi1024) [reply cupy/cuda/cufft.pyx]: Yes, please!

File: cupy/fft/config.py (hunk 8 above)
Comment (leofang) [reply cupy/fft/config.py]: Thanks, @asi1024! So if I understand you correctly, it is fine to keep `set_cufft_gpus()` as is. I suppose I better mark it experimental?

**Comment by leofang** [comment]
I hope I addressed all of your comments, @asi1024. PTAL.

File: cupy/fft/config.py (hunk 8 above)
Comment (asi1024) [reply cupy/fft/config.py]: Yes!

Comment (leofang) [reply cupy/fft/config.py]: I already did it. Thanks.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Comment by leofang** [comment]
Thanks a lot, @asi1024! I think the scatter/gather part might have room for optimization. Will look into that and create a follow up PR.
